compute_tools/src/compute.rs

```diff
@@ -316,6 +354,456 @@ pub(crate) fn construct_superuser_query(spec: &ComputeSpec) -> String {
+        let tokio_conn_conf = tokio_postgres::config::Config::from_str(connstr.as_str())
+            .context("cannot build tokio postgres config from connstr")?;
+
+        let pgversion = get_pg_version_string(&pgbin);
+
+        let mut new_state = ComputeState::new();
+        if let Some(cli_spec) = cli_spec {
+            let pspec = ParsedSpec::try_from(cli_spec).map_err(|msg| anyhow::anyhow!(msg))?;
+            new_state.pspec = Some(pspec);
+        }
+
+        Ok(ComputeNode {
+            compute_id: compute_id.to_string(),
+            connstr,
+            conn_conf,
+            tokio_conn_conf,
+            pgdata,
+            pgbin,
+            pgversion,
+            external_http_port,
+            internal_http_port,
+            live_config_allowed,
+            state: Mutex::new(new_state),
+            state_changed: Condvar::new(),
+            ext_remote_storage,
+            ext_download_progress: RwLock::new(HashMap::new()),
+            build_tag,
+            resize_swap_on_bind,
+            set_disk_quota_for_fs,
+            #[cfg(target_os = "linux")]
+            filecache_connstr,
+            #[cfg(target_os = "linux")]
+            cgroup,
+            #[cfg(target_os = "linux")]
+            vm_monitor_addr,
+        })
+    }
+
+    /// Top-level control flow of compute_ctl
+    pub fn run(self) -> Result<Option<i32>> {
+        let this = Arc::new(self);
+
+        let cli_spec = this.state.lock().unwrap().pspec.clone();
+
+        // If this is a pooled VM, prewarm before starting HTTP server and becoming
+        // available for binding. Prewarming helps Postgres start quicker later,
+        // because QEMU will already have its memory allocated from the host, and
+        // the necessary binaries will already be cached.
+        if cli_spec.is_none() {
+            this.prewarm_postgres()?;
+        }
+
+        // Launch the external HTTP server first, so that we can serve control plane
+        // requests while configuration is still in progress.
+        crate::http::server::Server::External(this.external_http_port).launch(&this);
+
+        // The internal HTTP server could be launched later, but there isn't much
+        // sense in waiting.
+        crate::http::server::Server::Internal(this.internal_http_port).launch(&this);
+
+        // If we got a spec from the CLI already, use that. Otherwise wait for the
+        // control plane to pass it to us with a /configure HTTP request
+        let pspec = if let Some(cli_spec) = cli_spec {
+            cli_spec
+        } else {
+            this.wait_spec()?
+        };
+
+        launch_lsn_lease_bg_task_for_static(&this);
+
+        // We have a spec, start the compute
+        let mut delay_exit = false;
+        let mut vm_monitor = None;
+        let mut pg_process: Option<PostgresHandle> = None;
+
+        let start_compute_result = this.start_compute(&mut pg_process);
+        if let Err(err) = start_compute_result {
```

Comment:
NIT: should it be just match? It's not clear why this extra var is needed

Reply:
converted to match